### PR TITLE
KAFKA-15770: IQv2 must return immutable position

### PR DESCRIPTION
**Pull request by mjsax**
ConsistencyVectorIntegrationTest failed frequently because the returned Position from IQv2 is not immutable while the test assumes immutability. To return a Position with a QueryResult that does not change, we need to deep copy the Position object.

Plus some side cleanup on the test itself.

### CODE REVIEW

**Comment by mjsax** [comment]
I am not 100% sure if my analysis is correct, but it's my best guess so far. Given that the test always passed locally, I would like to re-run Jenkins on this PR a few times (but in the end, I guess we need to merge optimistically, and maybe re-open the ticket afterward in case it's not fixed?)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java [hunk 1]
```diff
@@ -146,7 +146,7 @@ public static <R> QueryResult<R> handleBasicQueries(
                 "Handled in " + store.getClass() + " in " + (System.nanoTime() - start) + "ns"
             );
         }
-        result.setPosition(position);
+        result.setPosition(position.copy());
```
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java]: Nice find! But I'm not sure, whether this thread safe now? It seems I'd have to put a lock somewhere to make sure `position` isn't modified before I reach this point.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java]: Better yet, we could make the copy at the beginning of this method or even on the caller side.

**Review by vvcephei — APPROVED** [review]
Thanks, @mjsax ! This certainly looks like an improvement. I left a comment in case you want to tighten it down even more.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java]: Thanks. I was hoping it would be thready-safe already -- maybe not. Let me double check the code and figure it out.

**Comment by lucasbru** [comment]
Do we need to lock the implementations of `AbstractDualSchemaRocksDBSegmentedBytesStore.getWriteBatches` ?

**Comment by mjsax** [comment]
> Do we need to lock the implementations of AbstractDualSchemaRocksDBSegmentedBytesStore.getWriteBatches ?

Outstanding catch! I believe yes. Updated the PR.

File: streams/src/main/java/org/apache/kafka/streams/query/internals/SynchronizedPosition.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    public static SynchronizedPosition emptyPosition() {
+        return new SynchronizedPosition(new ConcurrentHashMap<>());
+    }
+
+    public static SynchronizedPosition fromMap(final Map<String, ? extends Map<Integer, Long>> map) {
+        return new SynchronizedPosition(deepCopy(map));
+    }
+
+    public static ConcurrentHashMap<String, ConcurrentHashMap<Integer, Long>> deepCopy(
```
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/query/internals/SynchronizedPosition.java]: Why did you move this static method? Since it's also accessed by `Position`, I'd rather leave it there.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractDualSchemaRocksDBSegmentedBytesStore.java [hunk 3]
```diff
@@ -193,17 +194,22 @@ public void put(final Bytes rawBaseKey,
...
-            StoreQueryUtils.updatePosition(position, stateStoreContext);
-
-            // Put to index first so that if put to base failed, when we iterate index, we will
-            // find no base value. If put to base first but putting to index fails, when we iterate
-            // index, we can't find the key but if we iterate over base store, we can find the key
-            // which lead to inconsistency.
-            if (hasIndex()) {
-                final KeyValue<Bytes, byte[]> indexKeyValue = getIndexKeyValue(rawBaseKey, value);
-                segment.put(indexKeyValue.key, indexKeyValue.value);
+            position.lock();
```
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractDualSchemaRocksDBSegmentedBytesStore.java]: Have you considered doing the same without `SynchronizedPosition` and instead just using the object monitor of position

```
synchronized(position) {


}
```
?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StoreQueryUtilsTest.java [hunk 4]
```diff
@@ -70,7 +71,7 @@ public void shouldReturnErrorOnBoundViolation() {
             PositionBound.at(Position.emptyPosition().withComponent("topic", 0, 1)),
             new QueryConfig(false),
             store,
-            Position.emptyPosition().withComponent("topic", 0, 0),
+            (SynchronizedPosition) Position.emptyPosition().withComponent("topic", 0, 0),
```
Comment (lucasbru) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StoreQueryUtilsTest.java]: That cast won't work

**Review by lucasbru — COMMENTED** [review]
I left a couple of comments. It also seems like there are a lot of test failures. But the general approach looks good to me, modulo non-snapshotted iterators which we will not fix in this PR.

**Comment by mjsax** [comment]
@lucasbru Updates this PR. -- When I started to work on this PR, I thought that using `synchronized` would not work, because locking/unlocking would not always be method-local. But you are right, turns out it is method local, so I dropped `SynchronizedPosition`, what makes your other comments obsolete.

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!
